### ISSUE TITLE
PS OOM in Distributed Inception

### ISSUE THREAD

**Issue by mbz** [tensorflow/tensorflow#6508]
We are trying to train Inception model on ImageNet in a distributed setting following the guide [here](https://github.com/tensorflow/models/tree/master/inception). TensorFlow successfully trains the model with 32 workers but with 64 worker, the PS dies with OOM killer. 

In all of these experiments, PS was running on a separate machine. Looking at the memory consumption of 32 workers, it seems that there is no memory issue and the memory becomes stable after a couple of iterations. With 64 workers, PS dies after allocating all the available memory (32GB) after only two iterations.

Is this the expected behavior from PS or not? If not, please let us know if we can provide any other information that may help debugging this problem.

#### Environment info
TensorFlow 0.12
CUDA 8.0
CUDNN 5.1
RAM: 32 GB
GPU: K20

**Comment by yaroslavvb** [comment]
How many PS tasks do you have? I would first check CPU utilization/network utilization of PS to make sure you have enough tasks to keep up with the workers. IE, suppose each worker can do a single batch in 2 seconds. With 25M parameters that's 50MB per second per worker, so in a shared cloud one might use 10 PS tasks for 32 workers and 20 PS tasks for 64 workers.

More generally, you can use Heap Checker from google profiling tools to get a break-down of what is taking up all the memory

**Comment by mbz** [comment]
Bandwidth shouldn't be an issue in this case since we have a 9.6GB/sec peak injection bandwidth to each node. For 32 workers, the current speed is less than 2 samples per second which should sum up to 3.2GB per second. 

If this helps, we are using the XK nodes in BlueWaters supercomputer (you can find the spec [here](https://bluewaters.ncsa.illinois.edu/hardware-summary)).

I will try using Heap Checker and hopefully will comeback with mode info.

**Comment by yaroslavvb** [comment]
What about CPU utilization of your PS task?

**Comment by mbz** [comment]
That can be the case. CPU utilization is at %100 but only one core. It seems to be single threaded. Is there any flag to use more threads?

Will send you the results of slow gradients as soon as I can.

**Comment by yaroslavvb** [comment]
More threads can be done with tf.Server(config arg, see #4455

**Comment by mbz** [comment]
I assume you are referring to `inter_op_parallelism_threads` and `intra_op_parallelism_threads`. Aren't these set to the number of available cores by default?

**Comment by yaroslavvb** [comment]
They should, although there have been cases when number of cores was
misidentified on new architecture

**Comment by yaroslavvb** [comment]
@mbz possibly related issue is https://github.com/tensorflow/tensorflow/issues/6116  Current grpc implementation uses too much CPU, there's a patch that fixes the issue

**Comment by mrry** [comment]
@mbz There are some guidelines on how to deal with memory leaks in TensorFlow here:

http://stackoverflow.com/documentation/tensorflow/3883/how-to-debug-a-memory-leak-in-tensorflow#t=201612281727521481272

In particular, you should try running with tcmalloc, which can prevent heap fragmentation (that sometimes leads to out-of-memory errors with the default `malloc()` implementation).

**Comment by mbz** [comment]
@yaroslavvb thanks for all the info. 

adding `tf.ConfigProto(inter_op_parallelism_threads=64,intra_op_parallelism_threads=64)` to `tf.train.Server` solved the issue. CPU utilization is way higher and the memory is stable at ~20GB peaking at 30GB. 

I can dig more into why TF does not pick up the right number of cores if you folks are interested. It would be helpful if you point me to the right code though.

@mrry this doesn't look like to be a memory leak. it's just high memory consumption due large number of received messages. By adding more threads to PS process (making processing faster) mitigates the memory issue, at least for 32 workers. The memory consumption still seems high though. Probably, because of gRPC overhead?

**Comment by mrry** [comment]
@mbz If you are not running with tcmalloc, I'd still recommend that you give it a try. Allocating and then soon after freeing a large number of large message objects may still be causing heap fragmentation.

**Comment by yaroslavvb** [comment]
@mbz You could check what `port::NumSchedulableCPUs` returns on your system. I see it being used in [local_device.cc](https://github.com/tensorflow/tensorflow/blob/d44d271c9da4d244ce4b2ffaf808adbe4cff759d/tensorflow/core/common_runtime/local_device.cc#L37)  to set number of threads

**Comment by mbz** [comment]
@mrry this graph might be interesting. this is the memory of PS overtime (one hour of log).

![image](https://cloud.githubusercontent.com/assets/1122127/21534139/147af5de-cd28-11e6-8865-9c04686b98d2.png)

which is in sync with network:

![image](https://cloud.githubusercontent.com/assets/1122127/21534144/24ebf44a-cd28-11e6-8539-ee27de20e32f.png)

unfortunately, it seems that Blue Waters (Cray to be more accurate) has some issues with `tcmalloc`. From [here](https://bluewaters.ncsa.illinois.edu/known-issues):
`TCMALLOC: Codes may crash in tcmalloc. Cray is working on fixing these. In the mean time, use the flag "-hsystem_alloc" to avoid linking in tcmalloc. Using this flag may result in a new set of problems. So use it with caution.`
I will still try it though, hope for the best.

**Comment by yaroslavvb** [comment]
To summarize, it looks like parameter server on Cray starts configured to use 1 core, which is not enough to keep up with 64 workers. As a result, the memory usage of ps grows without bound, possibly due to grpc backlog. The growth without bound might be something that can only be fixed on grpc side -- https://github.com/grpc/grpc

As far as TensorFlow thinking Cray computers only have one core, that could be filed as a separate issue.

@mbz Is there anything else to do on this issue before it's closed?

**Comment by mbz** [comment]
I believed that summarized it nicely. Thanks.
